modules/rostests/apitests/utildll/CachedGetUserFromSid.c

```diff
@@ -0,0 +1,102 @@
+    ULONG cchUserName;
+
+    // First, full success case, to load result into cache.
+    // Otherwise, expect misbehavior/corruption/crash!
+    // Same issues with 'NULL' or '0' arguments.
+    // Behavior changed on NT6.0 then NT6.1+...
+
+    wmemset(szUserName, L'?', _countof(szUserName));
+    cchUserName = _countof(szUserName);
+    pCachedGetUserFromSid(NULL, szUserName, &cchUserName);
+    ok(cchUserName == cchUserNameFromNull, "cchUserName: expected %lu, got %lu\n", cchUserNameFromNull, cchUserName);
+    ok(wcscmp(szUserName, szUserNameFromNull) == 0, "szUserName: expected \"%S\", got \"%.*S\"\n", szUserNameFromNull, (int)cchUserName, szUserName);
+
+    wmemset(szUserName, L'?', _countof(szUserName));
+    cchUserName = 1;
+    pCachedGetUserFromSid(NULL, szUserName, &cchUserName);
+    ok(cchUserName == 0, "cchUserName: expected 0, got %lu\n", cchUserName);
+    ok(szUserName[0] == UNICODE_NULL, "szUserName: missing UNICODE_NULL, got \"%.*S\"\n", (int)cchUserName, szUserName);
+
+    wmemset(szUserName, L'?', _countof(szUserName));
+    cchUserName = 2;
+    pCachedGetUserFromSid(NULL, szUserName, &cchUserName);
+    ok(cchUserName == 1, "cchUserName: expected 0, got %lu\n", cchUserName);
+    ok(szUserName[1] == UNICODE_NULL, "szUserName: missing UNICODE_NULL, got \"%.*S\"\n", (int)cchUserName, szUserName);
+    ok(wcsncmp(szUserName, szUserNameFromNull, 1) == 0, "szUserName: expected \"%.*S\", got \"%.*S\"\n", 1, szUserNameFromNull, (int)cchUserName, szUserName);
+
+    wmemset(szUserName, L'?', _countof(szUserName));
+    cchUserName = cchUserNameFromNull;
+    pCachedGetUserFromSid(NULL, szUserName, &cchUserName);
+    ok(cchUserName == cchUserNameFromNull - 1, "cchUserName: expected %lu, got %lu\n", cchUserNameFromNull - 1, cchUserName);
+    ok(szUserName[cchUserNameFromNull - 1] == UNICODE_NULL, "szUserName: missing UNICODE_NULL, got \"%.*S\"\n", (int)cchUserName, szUserName);
+    ok(wcsncmp(szUserName, szUserNameFromNull, cchUserNameFromNull - 1) == 0, "szUserName: expected \"%.*S\", got \"%.*S\"\n", (int)cchUserNameFromNull - 1, szUserNameFromNull, (int)cchUserName, szUserName);
+
+    wmemset(szUserName, L'?', _countof(szUserName));
+    cchUserName = cchUserNameFromNull + 1;
+    pCachedGetUserFromSid(NULL, szUserName, &cchUserName);
+    ok(cchUserName == cchUserNameFromNull, "cchUserName: expected %lu, got %lu\n", cchUserNameFromNull, cchUserName);
+    ok(wcscmp(szUserName, szUserNameFromNull) == 0, "szUserName: expected \"%S\", got \"%.*S\"\n", szUserNameFromNull, (int)cchUserName, szUserName);
+}
+
+START_TEST(CachedGetUserFromSid)
+{
+    // TODO: Dynamically checking, until ReactOS implements this dll.
+    const char szDll[] = "utildll.dll";
+    const char szFunction[] = "CachedGetUserFromSid";
+
+    HMODULE hModule = LoadLibraryA(szDll);
+    DWORD dwLE = GetLastError();
+
+    if (!hModule)
+    {
+        ok(FALSE, "LoadLibraryA(\"%s\") failed!\n", szDll);
+
+        if (dwLE != ERROR_MOD_NOT_FOUND)
```

Comment:
Printing the GLE code in case of error seems useful; testing it does not

Reply:
Merged.